### PR TITLE
Fixes for mode=preserve

### LINKED ISSUE #39279: when I use copy module, I set mode=preserve and remote_src=yes, It error
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
copy module
##### ANSIBLE VERSION
2.5.0
##### OS / ENVIRONMENT
centos 7.4
##### SUMMARY
when I use copy module, I set mode=preserve and remote_src=yes, It error:
`bad symbolic permission for mode: preserve`

##### STEPS TO REPRODUCE
```yaml
- name: copy test
  copy: src=/etc/gshadow dest=/destination owner=root group=root mode=preserve
```

##### ACTUAL RESULTS
```
bad symbolic permission for mode: preserve
```
##### Other
I found this [copy module should preserve mode of src file when mode is not explicitly se](https://github.com/ansible/ansible-modules-core/issues/1124). In this issue, nobody encountered the same problem as me. 
@ansibot can you fix this bug?

### PR DESCRIPTION
**Pull request by abadger**
* Document mode=preserve for template and copy
* Make mode=preserve work with remote_src for copy
* Make mode=preserve work for template

Fixes #39279

##### ISSUE TYPE

 - Bugfix Pull Request
 - Docs Pull Request

##### COMPONENT NAME
copy and template

##### ANSIBLE VERSION

```
devel
```

##### ADDITIONAL INFORMATION
~This just needs some integration tests to go along with mode=preserve for copy's remote_src=True and template.~

Tests added.  This is ready.

### CODE REVIEW

File: lib/ansible/modules/files/copy.py [hunk 1]
```diff
@@ -57,6 +57,17 @@
...
+  mode:
+    description:
+      - "Mode the file or directory should be. For those used to I(/usr/bin/chmod) remember that
+        modes are actually octal numbers.  You must either specify the leading zero so that
+        Ansible's YAML parser knows it is an octal number (like C(0644) or C(01777)) or quote it
+        (like C('644') or C('0644') so Ansible receives a string and can do its own conversion from
+        string into number.  Giving Ansible a number without following one of these rules will end
+        up with a decimal number which will have unexpected results.  As of version 1.8, the mode
+        may be specified as a symbolic mode (for example, C(u+rwx) or C(u=rw,g=r,o=r)).  As of
+        version 2.3, the mode may also be the special string C(preserve).  C(preserve) means that
```
Comment (jborean93) [inline lib/ansible/modules/files/copy.py:69]: Need to mention preserve only worked from local to remove from 2.3, 2.6+ would work with remote_src True

File: lib/ansible/modules/files/template.py [hunk 2]
```diff
@@ -95,6 +95,17 @@
...
+  mode:
+    description:
+      - "Mode the file or directory should be. For those used to I(/usr/bin/chmod) remember that
+        modes are actually octal numbers.  You must either specify the leading zero so that
+        Ansible's YAML parser knows it is an octal number (like C(0644) or C(01777)) or quote it
+        (like C('644') or C('0644') so Ansible receives a string and can do its own conversion from
+        string into number.  Giving Ansible a number without following one of these rules will end
+        up with a decimal number which will have unexpected results.  As of version 1.8, the mode
+        may be specified as a symbolic mode (for example, C(u+rwx) or C(u=rw,g=r,o=r)).  As of
+        version 2.3, the mode may also be the special string C(preserve).  C(preserve) means that
```
Comment (jborean93) [inline lib/ansible/modules/files/template.py]: Only copy with remote_src: no worked in 2.3, template would be 2.6+ unless you plan on backporting

**Review by jborean93 — COMMENTED** [review]
Just some minor comments on the description version added

**Review by jborean93 — APPROVED** [review]

File: lib/ansible/modules/files/copy.py (hunk 1 above)
Comment (abadger) [reply lib/ansible/modules/files/copy.py:69]: Fixed via a note in the remote_src description (which lists other remote_src limitations)

File: lib/ansible/modules/files/template.py (hunk 2 above)
Comment (abadger) [reply lib/ansible/modules/files/template.py]: Fixed in latest commit.

**Comment by abadger** [comment]
Thanks for fixing this @jborean93 !

The unstable tests are unrelated to this.  Merging to devel for the 2.6.0 release.

**Pull request merged**